File: java-to-kotlin-book/jtk-book/src/main/java/classes/EmailAddress.java
package classes;

import java.util.Objects;

public class EmailAddress {
	private final String localPart; // <1> - 불변 값 선언
	private final String domain;

	public static EmailAddress parse(String value) { // <2> - 정적 팩터리 메서드
		var atIndex = value.lastIndexOf('@');
		if (atIndex < 1 || atIndex == value.length() - 1)
			throw new IllegalArgumentException(
				"EmailAddress must be two parts separated by @"
			);
		return new EmailAddress(
			value.substring(0, atIndex),
			value.substring(atIndex + 1)
		);
	}

	public EmailAddress(String localPart, String domain) { // <3> 생성자 초기화
		this.localPart = localPart;
		this.domain = domain;
	}

	public String getLocalPart() { // <4> getter
		return localPart;
	}

	public String getDomain() { // <4> getter
		return domain;
	}

	@Override
	public boolean equals(Object o) { // <5> equals and hashcode 재정의
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		EmailAddress that = (EmailAddress) o;
		return localPart.equals(that.localPart) &&
			domain.equals(that.domain);
	}

	@Override
	public int hashCode() {
		return Objects.hash(localPart, domain);
	}

	@Override
	public String toString() { // <6> toString 재정의
		return localPart + "@" + domain;
	}
}

File: java-to-kotlin-book/jtk-book/src/main/kotlin/classes/EmailAddress_.kt
package classes

import java.util.*

/**
 * data class : equals, hashcode, toString 을 자동으로 생성해준다.
 *
 * 단, 데이터 클래스는 캡슐화를 제공하지 않는다. (copy method도 함께 생성하기 때문)
 * ex. val postmasterEmail = customerEmail.copy(localPart = "postmaster")
 * 위 예시처럼 내부 상테에 직접 접근하도록 허용해서 불변 조건을 깰 수 있다는 문제가 있다.
 */
data class EmailAddress_(
    val localPart: String,
    val domain: String
) {
    override fun toString(): String {
        return "$localPart@$domain"
    }

    companion object {
        @JvmStatic
        fun parse(value: String): EmailAddress_ {
            val atIndex = value.lastIndexOf('@')
            require(!(atIndex < 1 || atIndex == value.length - 1)) {
                "EmailAddress must be two parts separated by @"
            }
            return EmailAddress_(
                value.substring(0, atIndex),
                value.substring(atIndex + 1)
            )
        }
    }
}